File: src/services/pitStopDetector.ts
import type { PitStop, TelemetryPoint } from '../domain/models/telemetry';
import { FuelEngine } from './fuelEngine';

const CATEGORY_TAGS: Array<{
  category: PitStop['category'];
  name: string;
  keywords: string[];
}> = [
  { category: 'COFFEE_SHOP', name: 'Starbucks / Coffee Haven', keywords: ['coffee', 'cafe', 'bistro'] },
  { category: 'GAS_STATION', name: 'Shell / Energy Station', keywords: ['fuel', 'gas', 'shell', 'bp'] },
  { category: 'REST_AREA', name: 'Highway Rest Stop', keywords: ['rest', 'highway', 'plaza'] },
  { category: 'CLIENT_SITE', name: 'Downtown Tech Hub', keywords: ['office', 'tech', 'building'] },
  { category: 'PARKING', name: 'Central Parking Garage', keywords: ['parking', 'garage', 'lot'] },
  { category: 'TRAFFIC_HOLD', name: 'Severe Traffic Jam', keywords: ['traffic', 'jam', 'signal'] },
];

export class PitStopDetector {
  private activeStop: Partial<PitStop> | null = null;
  private fuelEngine: FuelEngine;

  constructor(fuelEngine: FuelEngine) {
    this.fuelEngine = fuelEngine;
  }

  /**
   * Process state change and point to detect and manage PitStops.
   */
  public evaluatePoint(
    point: TelemetryPoint,
    previousState: string
  ): PitStop | null {
    const isStopState = point.state === 'PIT_STOP' || point.state === 'IDLING';

    if (isStopState) {
      if (!this.activeStop) {
        // Initialize a new Pit-Stop cluster candidate
        const categoryMatch = this.predictCategory(point);
        this.activeStop = {
          id: `stop-${Date.now()}`,
          tripId: point.tripId,
          startTime: point.timestamp,
          latitude: point.latitude,
          longitude: point.longitude,
          category: categoryMatch.category,
          name: categoryMatch.name,
          address: `${point.latitude.toFixed(4)}° N, ${point.longitude.toFixed(4)}° W`,
          idleFuelCost: 0,
          idleFuelLiters: 0,
          durationSeconds: 0,
        };
      } else {
        // Update ongoing stop duration & fuel cost
        const durationSec = Math.round((point.timestamp - (this.activeStop.startTime || point.timestamp)) / 1000);
        const fuelWaste = this.fuelEngine.calculateIdleWaste(durationSec);

        this.activeStop.durationSeconds = durationSec;
        this.activeStop.idleFuelLiters = fuelWaste.idleLiters;
        this.activeStop.idleFuelCost = fuelWaste.idleCost;
      }
    } else if (previousState === 'PIT_STOP' || previousState === 'IDLING') {
      // Vehicle resumed movement -> Close and finalize pit stop
      if (this.activeStop && (this.activeStop.durationSeconds || 0) >= 30) {
        const completedStop: PitStop = {
          ...(this.activeStop as PitStop),
          endTime: point.timestamp,
        };
        this.activeStop = null;
        return completedStop;
      }
      this.activeStop = null;
    }

    return null;
  }

  public getActiveStop(): Partial<PitStop> | null {
    return this.activeStop;
  }

  public predictCategory(point: TelemetryPoint): { category: PitStop['category']; name: string } {
    // Spatial deterministic pseudo-picker for realistic demo stopping tags based on coords
    const latHash = Math.abs(Math.floor(point.latitude * 1000)) % CATEGORY_TAGS.length;
    const item = CATEGORY_TAGS[latHash] || CATEGORY_TAGS[0];
    return {
      category: item.category,
      name: item.name,
    };
  }

  public reset(): void {
    this.activeStop = null;
  }
}

File: smart_trip_ai_mobile/lib/services/pit_stop_detector.dart
import '../domain/models/telemetry.dart';
import 'fuel_engine.dart';

class CategoryTag {
  final PitStopCategory category;
  final String name;
  final List<String> keywords;

  CategoryTag(this.category, this.name, this.keywords);
}

final List<CategoryTag> categoryTags = [
  CategoryTag(PitStopCategory.coffeeShop, 'Starbucks / Coffee Haven', ['coffee', 'cafe', 'bistro']),
  CategoryTag(PitStopCategory.gasStation, 'Shell / Energy Station', ['fuel', 'gas', 'shell', 'bp']),
  CategoryTag(PitStopCategory.restArea, 'Highway Rest Stop', ['rest', 'highway', 'plaza']),
  CategoryTag(PitStopCategory.clientSite, 'Downtown Tech Hub', ['office', 'tech', 'building']),
  CategoryTag(PitStopCategory.parking, 'Central Parking Garage', ['parking', 'garage', 'lot']),
  CategoryTag(PitStopCategory.trafficHold, 'Severe Traffic Jam', ['traffic', 'jam', 'signal']),
];

class PitStopDetector {
  PitStop? _activeStop;
  final FuelEngine _fuelEngine;

  PitStopDetector(this._fuelEngine);

  /// Process state change and point to detect and manage PitStops.
  PitStop? evaluatePoint(TelemetryPoint point, String previousState) {
    final isStopState = point.state == DrivingState.pitStop || point.state == DrivingState.idling;

    if (isStopState) {
      if (_activeStop == null) {
        // Initialize a new Pit-Stop cluster candidate
        final categoryMatch = predictCategory(point);
        _activeStop = PitStop(
          id: 'stop-${DateTime.now().millisecondsSinceEpoch}',
          tripId: point.tripId,
          startTime: point.timestamp,
          latitude: point.latitude,
          longitude: point.longitude,
          category: categoryMatch.category,
          name: categoryMatch.name,
          address: '${point.latitude.toStringAsFixed(4)}° N, ${point.longitude.toStringAsFixed(4)}° W',
          idleFuelCost: 0,
          idleFuelLiters: 0,
          durationSeconds: 0,
        );
      } else {
        // Update ongoing stop duration & fuel cost
        final startTime = _activeStop!.startTime;
        final durationSec = ((point.timestamp - startTime) / 1000).round();
        final fuelWaste = _fuelEngine.calculateIdleWaste(durationSec);

        _activeStop = PitStop(
          id: _activeStop!.id,
          tripId: _activeStop!.tripId,
          startTime: _activeStop!.startTime,
          latitude: _activeStop!.latitude,
          longitude: _activeStop!.longitude,
          category: _activeStop!.category,
          name: _activeStop!.name,
          address: _activeStop!.address,
          durationSeconds: durationSec,
          idleFuelLiters: fuelWaste.idleLiters,
          idleFuelCost: fuelWaste.idleCost,
        );
      }
    } else if (previousState == 'PIT_STOP' || previousState == 'IDLING') {
      // Vehicle resumed movement -> Close and finalize pit stop
      if (_activeStop != null && _activeStop!.durationSeconds >= 30) {
        final completedStop = PitStop(
          id: _activeStop!.id,
          tripId: _activeStop!.tripId,
          startTime: _activeStop!.startTime,
          endTime: point.timestamp,
          latitude: _activeStop!.latitude,
          longitude: _activeStop!.longitude,
          category: _activeStop!.category,
          name: _activeStop!.name,
          address: _activeStop!.address,
          durationSeconds: _activeStop!.durationSeconds,
          idleFuelLiters: _activeStop!.idleFuelLiters,
          idleFuelCost: _activeStop!.idleFuelCost,
        );
        _activeStop = null;
        return completedStop;
      }
      _activeStop = null;
    }

    return null;
  }

  PitStop? getActiveStop() {
    return _activeStop;
  }

  CategoryTag predictCategory(TelemetryPoint point) {
    // Spatial deterministic pseudo-picker for realistic demo stopping tags based on coords
    final latHash = (point.latitude * 1000).floor().abs() % categoryTags.length;
    return categoryTags[latHash];
  }

  void reset() {
    _activeStop = null;
  }
}
